### DISCUSSION TITLE
Request Expiration Timestamp

### DISCUSSION THREAD

**Discussion by tsengia** [ankraft/ACME-oneM2M-CSE#104] [Q&A]
Just want to double check my thinking here:

The Request Expiration Timestamp (RET) is an absolute timestamp, correct?   

That means that if I want my request to timeout in 5 seconds, I need to send an RET that is equal to the current timestamp plus five seconds? Or can I just send an RET with a value equal to 5 seconds?

Thank you!

**Comment by reinaortega** [comment]
It can be either an absolute timestamp or a relative timestamp. So you can  send just RET equal to 5 seconds (I think it should be in ms)

**Reply by ankraft** [reply]
As Miguel wrote, it could be either:

- an absolute timestamp: In this case it must be an ISO8901 time string (UTC based, e.g. 20221115T104605), or ,
- a relative timestamp: In this case it is <del>either</del>:
    - <del>an ISO8901 period string (e.g. PT5S), or</del>
    - a number of milliseconds (e.g. 5000)

**Reply by tsengia** [reply]
To use this, I would add the `X-M2M-RET: 5000` header to my HTTP request, correct? which would give me a 5 second timeout?

**Reply by tsengia** [reply]
and ACME should close the connection when the timeout is reached?

**Reply by ankraft** [reply]
Yes, it should. It could be, however, under certain circumstances, for example if the request is currently processed by the CSE, that particular process is not hard interrupted. In general, the timeout test is done before the processing starts. 

The connection is not closed immediately, but you will get a response that the timeout has been reached. It is not the same as a timeout on the transport layer.

**Reply by tsengia** [reply]
Thank you, this timeout helped resolve a bug occurring on the Thingy:91. 

It turns out that the socket libraries authored by Nordic are missing the implementation of the `shutdown()` method, which is responsible for terminating the socket connection when the Thingy:91's request timeout it's reached. 

By setting this RET field, it is now ACME that responds to and terminates the connection, which allows our Thingy:91 to behave properly.

**Reply by ankraft** [reply]
This is great! I am sure that this took a lot of efforts to debug, and I am very glad about the outcome! 

Please put this in your report.

**Comment by reinaortega** [comment]
I think that for relative timestamp you have to use just a number. The ISO8901 string is for the xs:duration data type.

**Comment by ankraft** [comment]
Ah, you are right. I edited my comment.
